### PR TITLE
[Clang] Implement __reference_converts_from_temporary

### PR DESCRIPTION
**Pull request by cor3ntin**
This completes the required language support for P2255R2.

### CODE REVIEW

**Review by erichkeane — COMMENTED** [review]
Seems reasonable to me.  Curious about 1 change though.

**Comment by zyn0217** [comment]
@cjdb said he had discarded the implementation of this intrinsic in https://reviews.llvm.org/D135341 because he thought this was not as trivial as the `construct` version. Given that we have lost all the Phab histories, I think it'd be better to invite @cjdb to take a look at this? And I'm also curious what the previous implementation was :)

File: clang/lib/Parse/ParseDeclCXX.cpp [hunk 1]
```diff
@@ -1779,9 +1779,8 @@ void Parser::ParseClassSpecifier(tok::TokenKind TagTokKind,
           tok::kw___is_union,
           tok::kw___is_unsigned,
           tok::kw___is_void,
-          tok::kw___is_volatile,
-          tok::kw___reference_binds_to_temporary,
-          tok::kw___reference_constructs_from_temporary))
+          tok::kw___is_volatile
```
[the remark this answers is not in the document]
Comment (cor3ntin) [reply clang/lib/Parse/ParseDeclCXX.cpp:1782]: As they are new names, they were never structs. So there is no need for compatibility for old standard libraries.

File: clang/lib/Sema/SemaExprCXX.cpp [hunk 2]
```diff
@@ -5627,6 +5627,77 @@ static bool EvaluateUnaryTypeTrait(Sema &Self, TypeTrait UTT,
...
+  //   is_convertible<From, To> shall be satisfied if and only if
+  //   the return expression in the following code would be
+  //   well-formed, including any implicit conversions to the return
+  //   type of the function:
+  //
+  //     To test() {
+  //       return create<From>();
+  //     }
+  //
+  //   Access checking is performed as if in a context unrelated to To and
```
Comment (erichkeane) [inline clang/lib/Sema/SemaExprCXX.cpp:5654]: Do you have a test for this access checking?  I would expect something conversion ops/ctors that are protected, but they are friends of eachother.

**Review by erichkeane — APPROVED** [review]

**Comment by cor3ntin** [comment]
This breaks the build, I'll push a fix shortly.

**Pull request merged**